FEAT: Add initial sql server backend support

costrouc opened the pull request:
This commit is a minimal commit to add SQL Server support to ibis and pass all tests. It has been tested locally to work.

cc: @xmnlab

pep8speaks commented:
Hello @costrouc! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`ibis/sql/mssql/client.py`](https://github.com/ibis-project/ibis/blob/a46e605be78ecfedbf122337f5ae66697d9a0c84/ibis/sql/mssql/client.py):

> [Line 3:1](https://github.com/ibis-project/ibis/blob/a46e605be78ecfedbf122337f5ae66697d9a0c84/ibis/sql/mssql/client.py#L3): [F401](https://duckduckgo.com/?q=pep8%20F401) 'warnings' imported but unused

xmnlab commented on ci/docker-compose.yml:
```diff
@@ -91,6 +99,7 @@ services:
     image: jwilder/dockerize
     command: |
       dockerize -wait tcp://omniscidb:6274
+                -waut tcp://mssql:1433
```
typo: `waut` -> `wait`

costrouc replied:
Fixed. Thanks for finding that!

xmnlab commented:
@costrouc thanks for working on that.

it seems you are not using git pre-commit hook ... you can install that using `make develop` command.

Also use pydocstyle --convention=numpy to check locally the docstring of the new backend.

something like `pydocstyle --convention=numpy ibis/sql/mssql`

I am working to improve the docstring and docstring checks ... and I hope to add that to CI in a near future :)

xmnlab commented:
to run `pydocstyle` command probably you will need to remove `match` parameter on setup.cfg inside pydocstyle section

xmnlab commented on ci/docker-compose.yml:
```diff
@@ -9,6 +9,15 @@ services:
+      - 1433:1433
+    environment:
+      ACCEPT_EULA: Y
+      SA_PASSWORD: Rootm3plz
```
```
ERROR: Unable to set system administrator password: Password validation failed. The password does not meet SQL Server password policy requirements because it is not complex enough. The password must be at least 8 characters long and contain characters from three of the following four sets: Uppercase letters, Lowercase letters, Base 10 digits, and Symbols..
```

I think the mssql server is not starting because you are not using any symbol for the password .. but maybe I am wrong.

but just in case, could you change the password to something like `Ibis_MSSQL_2017`

costrouc replied:
Done.

xmnlab commented on ci/.env:
```diff
@@ -10,6 +10,11 @@ IBIS_TEST_MYSQL_PORT=3306
 IBIS_TEST_MYSQL_DATABASE=ibis_testing
+IBIS_TEST_MSSQL_HOST=mssql
+IBIS_TEST_MSSQL_PORT=1433
+IBIS_TEST_MSSQL_USER=ibis
```
probably you should use `sa` for user here

costrouc replied:
Done.

xmnlab commented on ci/.env:
```diff
@@ -10,6 +10,11 @@ IBIS_TEST_MYSQL_PORT=3306
+IBIS_TEST_MSSQL_HOST=mssql
+IBIS_TEST_MSSQL_PORT=1433
+IBIS_TEST_MSSQL_USER=ibis
+IBIS_TEST_MSSQL_PASSWORD=ibis
```
could you change the password to `Ibis_MSSQL_2017` ?

costrouc replied:
Done.

xmnlab commented:
@costrouc probably you should add mssql to:

https://github.com/ibis-project/ibis/blob/master/ci/azure/linux.yml#L52

and

https://github.com/ibis-project/ibis/blob/master/ci/azure/linux.yml#L189

xmnlab commented on ibis/sql/mssql/api.py:
```diff
@@ -0,0 +1,125 @@
+    ----------
+    expr : ibis.expr.types.Expr
+        The ibis expression to compile
+    params : dict or None
```
I think here you can use `params : dict, optional`

costrouc replied:
Fixed.

xmnlab commented on ibis/sql/mssql/api.py:
```diff
@@ -0,0 +1,125 @@
+    Parameters
+    ----------
+    host : string, default 'localhost'
+    user : string, default None
```
instead of `default None`, change it to `optional`. eg: `user : string, optional`
the same for the other parameters.

costrouc replied:
Fixed.

xmnlab commented on ibis/sql/mssql/client.py:
```diff
@@ -0,0 +1,196 @@
+
+        Parameters
+        ----------
+        name : str, optional
```
As we are using `string`, replace `str` to `string`.
probably we need to create a document and define standards, 
but for now I think would be good to use `string`

costrouc replied:
Fixed.

costrouc commented:
I fixed better than half of the docstring issues. Now I belive it is up to creating some sql server tests

xmnlab commented:
@costrouc some more things to add:


- add the `mssql` marker here: https://github.com/ibis-project/ibis/blob/master/setup.cfg#L34
- add `mssql` here https://github.com/ibis-project/ibis/blob/master/ci/load-data.sh#L5
- add mssql table schema for mssql here https://github.com/ibis-project/ibis/tree/master/ci/schema

xmnlab commented:
it seems the tests is not running for mssql:

```
SKIPPED [1] /opt/conda/envs/ibis-env/lib/python3.7/site-packages/_pytest/doctest.py:421: unable to import module local('/ibis/ibis/sql/mssql/api.py')
SKIPPED [1] /opt/conda/envs/ibis-env/lib/python3.7/site-packages/_pytest/doctest.py:421: unable to import module local('/ibis/ibis/sql/mssql/client.py')
SKIPPED [1] /opt/conda/envs/ibis-env/lib/python3.7/site-packages/_pytest/doctest.py:421: unable to import module local('/ibis/ibis/sql/mssql/compiler.py')
SKIPPED [355] /ibis/ibis/tests/backends.py:140: Unable to access module ibis.mssql
```

xmnlab commented:
@costrouc let me know if I can help you in any way here :)

xmnlab commented on ibis/sql/mssql/client.py:
```diff
@@ -0,0 +1,234 @@
+            return self.table_expr_class(node)
+
+    def list_tables(self, like=None, database=None, schema=None):
+        """List tables avilable in current database."""
```
@costrouc could you add to the docstrings the parameters and the return?

costrouc replied:
added.

xmnlab commented on ibis/sql/mssql/client.py:
```diff
@@ -0,0 +1,234 @@
+
+        Parameters
+        ----------
+
```
could you add `query` to the docstrings parameters section?

costrouc replied:
added.

xmnlab commented on ibis/tests/all/test_aggregation.py:
```diff
@@ -54,6 +58,7 @@
             lambda t, where: t.double_col.mean(),
             lambda t, where: t.double_col.mean(),
             id='mean',
+            # marks=pytest.mark.xfail_backends((MSSQL,)),
```
could you remove this commented code?

costrouc replied:
removed.

xmnlab commented on ibis/tests/all/test_aggregation.py:
```diff
@@ -146,6 +151,7 @@ def test_aggregation(
                 .reset_index()
             ),
             id='group_concat',
+            marks=pytest.mark.skip_backends((MSSQL,)),
```
any reason to use `skip_backend` here instead of `xfail_backends`?

costrouc replied:
no reason at all. changed to xfail_backends

xmnlab commented on ibis/tests/all/test_generic.py:
```diff
@@ -66,6 +67,7 @@ def test_coalesce(backend, con, expr, expected):
         assert result == expected
 
 
+@pytest.mark.skip_backends((MSSQL,))
```
any reason to use skip_backend here instead of xfail_backends?

costrouc replied:
no reason at all. changed to xfail_backends

xmnlab commented on ibis/tests/all/test_numeric.py:
```diff
@@ -304,6 +305,7 @@ def test_binary_arithmetic_operations(backend, alltypes, df, op):
     )
 
 
+@pytest.mark.skip_backends((MSSQL,))
```
any reason to use `skip_backends` here instead of `xfail_backends`?

costrouc replied:
no reason at all. changed to xfail_backends

xmnlab commented on ibis/tests/all/test_numeric.py:
```diff
@@ -314,6 +316,7 @@ def test_mod(backend, alltypes, df):
     backend.assert_series_equal(result, expected, check_dtype=False)
 
 
+@pytest.mark.skip_backends((MSSQL,))
```
any reason to use `skip_backends` here instead of `xfail_backends`?

costrouc replied:
no reason at all. changed to xfail_backends

costrouc commented:
@xmnlab your requested changes have been made. Open for more feedback!

xmnlab commented on docs/source/release.rst:
```diff
@@ -7,6 +7,7 @@ Release Notes
    These release notes are for versions of ibis **1.0 and later**. Release
    notes for pre-1.0 versions of ibis can be found at :doc:`/release-pre-1.0`
 
+* :feature:`1997` Add initial sql server backend support
```
@costrouc could you move your release note just below `* :release:`1.2.1 <pending>`` ?

costrouc replied:
Moved. Thanks! Didn't want to ask you to include it in the 1.2.1 release but I appreciate that it will be

jreback commented:
yes let me have a look

costrouc commented:
@jreback just a gentle reminder. Have you had time to look at the PR? Hope that you had a great holiday.

xmnlab commented:
hey @jreback just a gentile reminder about this PR :)

jreback commented on ibis/sql/mssql/compiler.py:
```diff
@@ -0,0 +1,207 @@
+
+# Aggregation
+# coppied from postgresql compiler
+# support for of bit columns in aggregate methods
```
can you import directly from the postgresql compiler (and/or move it to a generic function)

costrouc replied:
I removed the comments since this method actually cannot use the postgresql method due to the sql function having a different name. But otherwise this method is very similar.

jreback commented on ibis/sql/mssql/compiler.py:
```diff
@@ -0,0 +1,207 @@
+
+
+# String
+# TODO: substr and find are copied from SQLite, we should really have a
```
isn't ibis/sql/* the place for this?

happy to refactor this later, though might be useful to do a pre-cursor where you factor out from the various sqlalchemy based backends

costrouc replied:
I removed the comments since this method actually cannot use the sqlite method due to the sql function having a different name. But otherwise this method is very similar. Looks like the section you are talking about is in `ibis/sql/*` were you commenting on a different file?

jreback replied:
after this PR it makes sense to refactor the sql backends to share common code - per backend names can easily be parameter used - copying code is not great - but as a follow up is ok

if u can create an issue

jreback commented:
@cpcloud if you could have a look :>

jreback commented:
@icexelloss @toryhaavik if you'd have a look.

xmnlab commented:
@costrouc could you rebase your could again?

xmnlab commented:
@jreback any new about the review process? it looks in a good shape .. and maybe if anyone else want to review that later we can address that in a follow up PR. what do you think?

jreback commented:
> @jreback any new about the review process? it looks in a good shape .. and maybe if anyone else want to review that later we can address that in a follow up PR. what do you think?

@xmnlab thats not how reviews work - best to look in the original PR

ping @icexelloss if u can have a look

toryhaavik commented:
I'll take a look in the next couple days

On Fri, Jan 31, 2020, 20:58 Jeff Reback <notifications@github.com> wrote:

> @jreback <https://github.com/jreback> any new about the review process?
> it looks in a good shape .. and maybe if anyone else want to review that
> later we can address that in a follow up PR. what do you think?
>
> @xmnlab <https://github.com/xmnlab> thats not how reviews work - best to
> look in the original PR
>
> ping @icexelloss <https://github.com/icexelloss> if u can have a look
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ibis-project/ibis/pull/1997?email_source=notifications&email_token=ABASCASLXTRTSLNSEUQZ3QLRATJMVA5CNFSM4JATRMH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKQQNZA#issuecomment-580978404>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABASCAWKCDZIB6R7HMRPOR3RATJMVANCNFSM4JATRMHQ>
> .
>

icexelloss commented on ibis/tests/all/test_string.py:
```diff
@@ -3,7 +3,14 @@
+    PostgreSQL,
+    PySpark,
+    Spark,
+    MSSQL,
```
Let's keep alphabetical import order

costrouc replied:
fixed by isort

icexelloss commented on setup.py:
```diff
@@ -20,6 +20,7 @@
 
 sqlite_requires = ['sqlalchemy>=1.1,<1.3.7']
 postgres_requires = sqlite_requires + ['psycopg2']
+mssql_requires = sqlite_requires + ['pyodbc']
```
In general, I wonder if there is a way to do optional dependencies? i.e., an ibis user that doesn't use it with mssql shouldn't need to install `pyodbc`?

costrouc replied:
This is exactly how this is already configured. To install mssql you would `pip install ibis[mssql]`

icexelloss commented on ibis/tests/all/test_numeric.py:
```diff
@@ -10,7 +10,7 @@
 import ibis
 from ibis import literal as L
 from ibis.expr import datatypes as dt
-from ibis.tests.backends import MySQL, OmniSciDB, PostgreSQL
```
Please fix import order

icexelloss commented on ibis/tests/all/test_numeric.py:
```diff
@@ -10,7 +10,7 @@
 from ibis import literal as L
 from ibis.expr import datatypes as dt
-from ibis.tests.backends import MySQL, OmniSciDB, PostgreSQL
+from ibis.tests.backends import MySQL, OmniSciDB, PostgreSQL, MSSQL
```
Please fix import order

costrouc replied:
fixed by isort

icexelloss commented on ibis/tests/all/test_generic.py:
```diff
@@ -8,6 +8,7 @@
     BigQuery,
     Clickhouse,
     MySQL,
+    MSSQL,
```
Please fix import order

costrouc replied:
fixed by isort

icexelloss commented on ibis/tests/all/test_aggregation.py:
```diff
@@ -2,7 +2,7 @@
 from pytest import param
 
-from ibis.tests.backends import Clickhouse, MySQL, OmniSciDB, SQLite
+from ibis.tests.backends import Clickhouse, MySQL, OmniSciDB, SQLite, MSSQL
```
Please fix import order

costrouc replied:
fixed by isort

icexelloss commented on ci/datamgr.py:
```diff
@@ -416,6 +416,36 @@ def omniscidb(schema, tables, data_directory, **params):
 
 
+@cli.command()
+@click.option('-h', '--host', default='localhost')
```
Can we avoid setting these configs in two places?

costrouc replied:
This is how all other backends implement it. Trying to make this PR focused on initial sql server support.

icexelloss commented on ci/datamgr.py:
```diff
@@ -416,6 +416,36 @@ def omniscidb(schema, tables, data_directory, **params):
+)
+@click.option('-t', '--tables', multiple=True, default=TEST_TABLES)
+@click.option('-d', '--data-directory', default=DATA_DIR)
+def mssql(schema, tables, data_directory, **params):
```
I am not very famliar with database set ups..Can you please explain how is this method used?

costrouc replied:
This method is used to initalized the test dataset for the CI for sql server

icexelloss commented on ibis/sql/mssql/compiler.py:
```diff
@@ -0,0 +1,200 @@
+    return reduction_compiler
+
+
+def _substr(t, expr):
```
At the high level, can you explain why mssql compiler here needs to implement custom rules instead of using the shared sql compiler?

costrouc replied:
This method was required and not using the shared sql compiler because of this line `    f = sa.func.substring` since sql server uses a different function name. So this shows that the general shared sql compiler is not general enough. I've seen other sql backends have to do the same thing (e.g. postgresql)

jreback commented:
@costrouc if you can merge master and fix conflicts.

@toryhaavik if you can have a look.

jreback commented:
can you merge master and we'll have another look at this

costrouc commented:
@jreback I have rebased this PR and make it compatible with the recent CI changes. I have been contacted by two people regarding further work on the sql server backend which depend on getting this initial PR merged. isort now enforces import orders so those issues mentioned above are resolved. In addition there was a comment above about why the `ibis/sql/compiler.py` given methods were not used. For each case where I implemented a method in `ibis/sql/mssql/compiler.py` the generic method was either not general enough or sql servers behavior did not match the type conversions that the generic method expects. This is ready for merge upon your approval.

costrouc commented:
The failed tests above have nothing to do with sql server tests. One is loading postgresql data for windows 37 and another is test_change_properties for impala.

jreback commented:
@toryhaavik @icexelloss if you'd have a look

jreback commented:
> The failed tests above have nothing to do with sql server tests. One is loading postgresql data for windows 37 and another is test_change_properties for impala.

did you rebase on master?

costrouc commented:
@jreback yes I rebased on master. Latest is `upstream/master FEAT: OmniSciDB IsNan implementation (#2093)`

costrouc commented:
Looks like master branch (upon which I rebased) is currently failing as well https://github.com/ibis-project/ibis/runs/540420023?

jreback commented:
@costrouc can you rebase again, just merged a patch.

costrouc commented:
@jreback I have rebased on the latest of master

costrouc commented:
After rebasing again still running into errors with `no space left on device`. Master seems to have this same problem

jreback commented:
> After rebasing again still running into errors with `no space left on device`. Master seems to have this same problem

I merged the patch yesterday, are you sure you are rebasing on HEAD?

costrouc commented:
@jreback yes. https://github.com/costrouc/ibis/commits/sql-server-support?after=421a22277b29fae687c14e3d4e03f7a4a8bcb3a0+34

costrouc commented:
The most recent commit to master for ibis has the same `no space left on device` issue. Here is a link to the CI run https://github.com/ibis-project/ibis/runs/573978990. Please can we get this PR merged. I have rebased many times in the past 4 months to adapt with changes in testing. I have gone through your review, @xmnlab, and @icexelloss. All sql server tests pass locally and on the CI evidenced by https://dev.azure.com/ibis-project/ibis/_build/results?buildId=2255&view=logs&jobId=bd1d7a9f-a285-5985-c638-12e37ef5ca2a. I would really like to extend sql server support past this PR but I am trying to keep this minimal to give it the best chance of being merged.

xmnlab commented:
I can check if we can refactor/optimize more the docker stuff, but my first guess is that we are already running it at its limit, as we had the same issue few days ago. Split jobs by services would be an option?

costrouc commented:
@jreback rebased and same issue only jobs that fail are the ones that run out of disk space.

jreback commented:
@costrouc the CI is still broken, though being worked on in: https://github.com/ibis-project/ibis/pull/2196

we cannot merge *anything* unless this is fixed.

jreback commented:
@costrouc 
ok we are finally back to green. However, you need to update the config add testing for this build. do not add to existing as these are already overloaded, please create a new back end build.

you can find the pattern here: https://github.com/ibis-project/ibis/commit/8910d40854a841b86c8e8509d41cf5d9c9109b88

ping when ready.

costrouc commented:
Should I make those changes now? It looks like master still is not passing https://github.com/ibis-project/ibis/commits/master.

jreback commented:
master is passing, yes please rebase, you can add this job to the sql ones (don't need to create a new one)

jreback commented:
@costrouc I think the CI should be pretty ok for this patch now. this would need a separate backend i think (see how we split out omnisc / spark), though if you can get it to run on the existing sql ones then this would be ok.

datapythonista commented:
Spoke with @costrouc. For now it's probably better to keep the MSSQL backend as a separate project. It can be easily installed with pip and conda, so there is probably not much benefit on moving it to this repo. I'll be adding it to the docs in #2281, so users are aware of it.
